libathemecore/atheme.c

```diff
@@ -183,6 +183,94 @@ detach_console(int *daemonize_pipe)
 #endif
 }
 
+static int ATHEME_FATTR_PRINTF(3, 4)
+test_vsnprintf_iso_c99_driver(char *const restrict buf, const size_t len, const char *const restrict fmt, ...)
+{
+	va_list ap;
+	va_start(ap, fmt);
+	const int result = vsnprintf(buf, len, fmt, ap);
+	va_end(ap);
+	return result;
+}
+
+static bool
+test_vsnprintf_iso_c99(void)
+{
+	const long double fourty_two_ld = 42.0L;
+	const size_t fourty_two_sz = 42U;
+	char buf[BUFSIZE];
+
+	/* To keep Valgrind etc happy (uninitialised memory access in strncmp(3) below?)
+	 * We also don't use null bytes, to verify that vsnprintf(3) writes those too.
+	 */
+	(void) memset(buf, 0xFF, sizeof buf);
+
+	/* C99 vsnprintf(3) should return the number of characters it /would've/ written, even with a short buffer.
+	 * This also tests whether C99 format token 'LF' (long double) and field width specifiers work correctly.
+	 */
+	if (test_vsnprintf_iso_c99_driver(buf, 14U, "The answer is %27.2LF.", fourty_two_ld) != 42)
+		return false;
+
+	/* It should also return that for a buffer size of zero; *and* not write to the buffer at all in this case.
+	 * This also tests whether C99 format token 'zu' (size_t) and field width specifiers work correctly.
+	 */
+	if (test_vsnprintf_iso_c99_driver(buf, 0U, "No, it's %32zu.", fourty_two_sz) != 42)
+		return false;
+
+	// Verify that the first time was truncated correctly, and that the second time didn't write at all.
+	if (strncmp(buf, "The answer is", 14U) != 0)
+		return false;
+
+	return true;
+}
+
+static bool
+test_snprintf_iso_c99(void)
+{
+	// This is the same as above but for snprintf(3) instead
+
+	const long double fourty_two_ld = 42.0L;
+	const size_t fourty_two_sz = 42U;
+	char buf[BUFSIZE];
+
+	(void) memset(buf, 0xFF, sizeof buf);
+
+	if (snprintf(buf, 14U, "The answer is %27.2LF.", fourty_two_ld) != 42)
+		return false;
+
+	if (snprintf(buf, 0U, "No, it's %32zu.", fourty_two_sz) != 42)
+		return false;
+
+	if (strncmp(buf, "The answer is", 14U) != 0)
+		return false;
+
+	return true;
+}
```

Comment:
1. Nice example text, I approve
2. Apologies if you already mentioned this internally; could we test for this at configure time instead? (That wouldn't catch cross-compilation troubles, though I suspect we already fail at that. Also, "yes but autoconf" is probably an acceptable answer.)

Reply:
No need to test at configure or runtime now that the configure script itself aborts if we're not using a C99 compiler. This test has been removed.